In [2]:
import numpy as np 
import pandas as pd

In [18]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv']))



In [19]:
df.shape

(139, 17)

In [20]:
df

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0   0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1   0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2   0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3   0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4   0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   
..       ...       ...        ...           ...       ...           ...   
1   0.256154  0.126923   0.093846      0.310769  0.493077      0.361538   
2   0.420769  0.230000   0.326923      0.450769  0.328462      0.130769   
0   0.116923  0.238462  -0.021538     -0.071538 -0.500769     -0.336923   
1  -0.103077  0.056154  -0.315385     -0.459231 -0.072308     -0.453846   
0   0.326154  0.253077   0.265385      0.240769 -0.143077     -0.050769   

    takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0    0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1    0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2    0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3    0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4    0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   
..        ...         ...        ...              ...           ...       ...   
1    0.312308    0.273077   0.276154         0.883077     -0.286923 -0.134615   
2    0.607692    0.508462   0.564615         0.351538     -0.436923 -0.271538   
0    0.068462    0.306923  -0.611538        -0.428462      0.516154 -0.138462   
1   -0.075385   -0.259231   0.113846         0.347692      0.386154 -0.102308   
0    0.340769    0.227692  -0.430769        -0.398462      0.040000 -0.013077   

    aerials_z   fouls_z  fouled_z    nsxg_z  results  
0   -0.020000 -0.224615  0.100769 -0.124615        2  
1    0.196923 -0.016923  0.033846  0.146923        1  
2    0.683846  0.213846  0.353846  0.786923        9  
3    0.038462  0.039231  0.022308  0.692308        5  
4    0.572308 -0.016154 -0.096923  0.890769        5  
..        ...       ...       ...       ...      ...  
1   -0.229231  0.307692 -0.050769  0.482308        7  
2   -0.109231  0.100000 -0.008462  0.846923        4  
0   -0.083077 -0.126154  0.086154 -0.119231       12  
1   -0.430000 -0.211538 -0.019231 -0.552308       28  
0   -0.039231  0.298462  0.220769  0.229231        3  

[139 rows x 17 columns]

In [21]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.600857   0.044892      0.276148  0.285426   
xg_z             0.600857  1.000000   0.352025      0.639377  0.406542   
crosses_z        0.044892  0.352025   1.000000      0.630704  0.259117   
boxtouches_z     0.276148  0.639377   0.630704      1.000000  0.419614   
passes_z         0.285426  0.406542   0.259117      0.419614  1.000000   
progpasses_z     0.178107  0.227415   0.347515      0.323921  0.727346   
takeons_z        0.233298  0.303336   0.064599      0.160554  0.429705   
progruns_z       0.205235  0.442038   0.265906      0.451447  0.673331   
tackles_z        0.320715  0.397802   0.186701      0.382981  0.755030   
interceptions_z  0.284272  0.343790   0.223793      0.341322  0.705450   
clearances_z    -0.157091 -0.336867  -0.181647     -0.418735 -0.191357   
blocks_z         0.143022  0.194553   0.129358      0.317094  0.390330   
aerials_z        0.071112  0.099873   0.240333      0.272195 -0.011988   
fouls_z         -0.136249 -0.120087   0.051685     -0.113696 -0.096172   
fouled_z        -0.029775  0.041424   0.137568      0.050345  0.244980   
nsxg_z           0.346813  0.647027   0.677477      0.870832  0.552416   
results         -0.648327 -0.351595   0.013193     -0.219873 -0.213634   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.178107   0.233298    0.205235   0.320715   
xg_z                 0.227415   0.303336    0.442038   0.397802   
crosses_z            0.347515   0.064599    0.265906   0.186701   
boxtouches_z         0.323921   0.160554    0.451447   0.382981   
passes_z             0.727346   0.429705    0.673331   0.755030   
progpasses_z         1.000000   0.204778    0.440982   0.514591   
takeons_z            0.204778   1.000000    0.594197   0.384164   
progruns_z           0.440982   0.594197    1.000000   0.605932   
tackles_z            0.514591   0.384164    0.605932   1.000000   
interceptions_z      0.499489   0.409978    0.508053   0.704717   
clearances_z        -0.069242  -0.004335   -0.103507  -0.246465   
blocks_z             0.327274   0.309399    0.404493   0.331527   
aerials_z            0.036117  -0.043491    0.026571   0.001531   
fouls_z              0.098379   0.086723    0.007555  -0.122438   
fouled_z             0.085221   0.353731    0.331936   0.247186   
nsxg_z               0.465986   0.308203    0.582989   0.496801   
results             -0.130977  -0.164761   -0.222986  -0.272291   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.284272     -0.157091  0.143022   0.071112 -0.136249   
xg_z                    0.343790     -0.336867  0.194553   0.099873 -0.120087   
crosses_z               0.223793     -0.181647  0.129358   0.240333  0.051685   
boxtouches_z            0.341322     -0.418735  0.317094   0.272195 -0.113696   
passes_z                0.705450     -0.191357  0.390330  -0.011988 -0.096172   
progpasses_z            0.499489     -0.069242  0.327274   0.036117  0.098379   
takeons_z               0.409978     -0.004335  0.309399  -0.043491  0.086723   
progruns_z              0.508053     -0.103507  0.404493   0.026571  0.007555   
tackles_z               0.704717     -0.246465  0.331527   0.001531 -0.122438   
interceptions_z         1.000000     -0.170258  0.281012   0.040112 -0.043143   
clearances_z           -0.170258      1.000000  0.097544   0.241418  0.307169   
blocks_z                0.281012      0.097544  1.000000   0.150074  0.116196   
aerials_z               0.040112      0.241418  0.150074   1.000000  0.262609   
fouls_z                -0.043143      0.307169  0.116196   0.262609  1.000000   
fouled_z                0.257478      0.201586  0.099460   0.030483  0.185005   
nsxg_z                  0.464000     -0.412382  0.317087   0.160535 -0.062909   
results                -0.179392      0.087176 -0.141556  -0.114751  0.167070   

                 fo

In [115]:
df.shape

(115, 17)

In [116]:
# df.to_csv('WC-stats.csv', index=False)


In [33]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [27]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [28]:
y = df['results']

In [29]:
X = df.drop(columns=['results'])

In [31]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [34]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.4494483686544971

In [35]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [36]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 0.982 +/- 0.123
    progruns_z0.132 +/- 0.039
    xg_z    0.056 +/- 0.017
    boxtouches_z0.046 +/- 0.023
    nsxg_z  0.039 +/- 0.013
    aerials_z0.031 +/- 0.014
neg_mean_absolute_percentage_error
    goals_z 0.888 +/- 0.122
    progruns_z0.181 +/- 0.059
neg_mean_squared_error
    goals_z 89.125 +/- 11.206
    progruns_z11.964 +/- 3.505
    xg_z    5.055 +/- 1.543
    boxtouches_z4.180 +/- 2.062
    nsxg_z  3.509 +/- 1.182
    aerials_z2.782 +/- 1.244


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.